docs/documentation/input-processing.md

```diff
@@ -0,0 +1,1942 @@
+frequently have higher security requirements as an LMS, like pen-tests, systematic
+training and instruction of developers, risk analysis, automated or manual code
+reviews, are not implemented in the general developement process of ILIAS. This
+on the one hand means that there certainly are low hanging fruits to be picked
+regarding security. On the other hand it again shows that technical approaches
+most certainly won't be enough to raise the general level of security. Even if
+considering social implications, like we attempt to do in this paper, the
+improvement of the security of input processing won't be enough in general and
+will have a hard stand if not backed by other means. There are, however, measures
+in the scope of input processing, like proper communication and good API-design,
+that might help to form new habbits regarding security in the long run.
+
+We thus try to show approaches besides technical measures for the scope of this
+paper, but also ask the project leadership of ILIAS to keep thinking about and
+implementing more hollistic approaches to security.
+
+
+### Feedback when Rejecting Input
+
+The circumstances in which the system receives input have a huge variety regarding
+their sources. There are forms that are operated by users, who need to be informed
+about success or failure of the desired input. There are types of input
+that are very technical in their nature, like reading data from the database or
+XML-import files, where machines are talking to other machines along a more or
+less strictly defined interface. There are forms of input that seem to be machine-
+to-machine like the former, but where we in fact don't really know if and how
+humans that need feedback are involved, e.g. the SOAP-Interface or JSON-over-HTTP-
+interfaces.
+
+The requirements regarding the mere filtering of the data that reach the system
+are very similar in all these cases: We need to make sure that only data matching
+certain criteria in shape and content may pass the boundary of the system and be
+subject to further processing. Data should be scrutinized as closely and early
+as possible. The requirements regarding feedback to the other side of the
+communication are, however, widely different.
+
+A human sitting in front of a form might just input improper data and every nice
+system will try to give her feedback in the most helpful and detailed form that's
+possible. Reading some data from the database seems to require no feedback to a
+user at all, as every missformed data indicates some deeper problem that most
+certainly cannot be solved at the interface to the database itself. A malformed
+XML-file in an import might hint at incompatible versions or objects regarding the
+import, as well as at some attempt to tamper with an import file and use it as a
+vector to degrade security of the system. A user importing the file might require
+the feedback "incompatible version", but less likely seems to be interested in
```

Comment:
"Incompatible version" is not an issue when using XML-import that I am aware of, since no functions have been abandoned in last years (besides MD5 encryption). On the other hand, user often have some wrong data in a huge import file, so I believe a detailed feedback on the exact field is absolutely necessary.

KR,
Stas

Reply:
Hi @swischniak,

thx a lot for your feedback.

This paragraph is supposed to be an example for different requirements on feedback when inputting data. It's not a stipulation on what is required when using XML-imports. I tried to clarify that by using various "might"s etc. Do you agree with the general conclusion that we should take care about outputting feedback with different levels of detail according to the situation where we output that feedback? I'll be happy to change my example, though. Do you have another example (even of a hypothetical situation) where a human user does not need a full detailed error report regarding his input and where a full report would unnecessarily disclose information to an attacker?

For my curiosity: I never encountered the situation of wrong data in an import file. ILIAS exports them, ILIAS imports them. What are your users doing to introduce wrong data in these files?

Best regards!

Reply:
Hi Richard @klees,

thx for your reply.
Frankly, I am not sure if we need different levels of feedback. When speaking of XML Import my experience is based on different interfaces (e.g. SOAP) and XML Import via Excel macro generating XML, which I am maintaining. So in my experience at the end it is always a human, an ILIAS-Administrator, who has to evaluate the feedback, analyse the mistake and make adjustments to the imports accordingly. So what I would expect when importing 3000 users is not a list on what went right (“Completed successfully” would be sufficient here) but some hints on the line (that’ll teach you to use a well-formed XML ;-) ) and e.g. on a value or syntax, ILIAS has problems with and on what kind of problem it was (value too large, wrong format, numeric expected, wrong attribute etc.).
And at this stage I am less worried about potential attackers, since you do need appropriate rights to even start importing and on the other hand some “obscurity” in feedback on wrong input will not stop you.

Said that, there of course might be other levels of feedback needed in features I am not considering, for example a verbose output in a log. But since “the logging-topic is considered to be out of scope for the rest of the paper” I am just not aware of any other scenarios, which should have different feedback levels.

And for your curiosity: No situation with wrong data? Export, than import – that easy, eh? Lucky you. :-)
When using interfaces, the data is often from a third party system and you usually can’t rely on it. Same with Excel Import. We try to check on some mistakes in excel already, but will be always many steps behind on that, considering what systems, formulas and hand the data has been thru. 

[You might also want to have a look in user import forum to have actual examples on that:
e.g. ["User file import failing"](https://docu.ilias.de/goto_docu_frm_4691_5333.html)

KR,
Stas

Reply:
Hi @swischniak,

thanks again for your feedback. I see your use-case for fine grained error output, but I still see, that different levels of output are required for different scenarios regarding security. The poster child for these different output levels imo is the devmode. We want fine grained output for devs on development instances but certainly not for users on live instances.

I hope to hear you tomorrow in the VC, maybe we can continue the discussion there.

Best regards!